include/flatbuffers/grpc.h

```diff
@@ -23,50 +23,232 @@
 #include "grpc++/support/byte_buffer.h"
 #include "grpc/byte_buffer_reader.h"
 
+namespace flatbuffers {
 namespace grpc {
 
+// Message is a typed wrapper around a buffer that manages the underlying
+// `grpc_slice` and also provides flatbuffers-specific helpers such as `Verify`
+// and `GetRoot`. Since it is backed by a `grpc_slice`, the underlying buffer
+// is refcounted and ownership is be managed automatically.
 template <class T>
-class SerializationTraits<T, typename std::enable_if<std::is_base_of<
-                                 flatbuffers::BufferRefBase, T>::value>::type> {
+class Message {
  public:
-  // The type we're passing here is a BufferRef, which is already serialized
-  // FlatBuffer data, which then gets passed to GRPC.
-  static grpc::Status Serialize(const T& msg,
-                                grpc_byte_buffer **buffer,
-                                bool *own_buffer) {
-    // TODO(wvo): make this work without copying.
-    auto slice = gpr_slice_from_copied_buffer(
-                   reinterpret_cast<const char *>(msg.buf), msg.len);
-    *buffer = grpc_raw_byte_buffer_create(&slice, 1);
-    grpc_slice_unref(slice);
+  Message() : slice_(grpc_empty_slice()) {}
+
+  Message(grpc_slice slice, bool add_ref)
+    : slice_(add_ref ? grpc_slice_ref(slice) : slice) {}
+
+#if FLATBUFFERS_GRPC_ENABLE_MESSAGE_COPY
```

Comment:
Actually if your point is that we should remove them entirely, I'm okay with that.

I don't know for sure if gRPC buffers are reused. If they are, then messages are not safe to keep outside of the rpc function, and there pretty much be no reason to copy/assign a `flatbuffers::grpc::Message<T>`. @ctiller, could you comment on the gRPC buffer reuse policy just so I know?

I think in either case I'll just remove these, since the semantics of a copy might be misleading if it's still tied to the same buffer.